designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,505 @@
+1. **Encourage plugins to use `configs`.** For plugin authors, we want to encourage the use of a consistent entrypoint, `configs`, where users can find predefined configurations. 
+
+## Detailed Design
+
+Design Summary:
+
+1. Allow arrays in config arrays (in addition to objects)
+1. Introduce an `extends` key in config objects
+
+### Allow Arrays in Config Arrays
+
+This was part of the original RFC but was disabled over concerns about the complexity involved with importing configs from packages and having no idea whether it was an object, an array, or an array with a mix of objects and arrays. This mattered most when users needed to use `map()` to apply different `files` or `ignores` to configs, but with `extends`, that concern goes away.
+
+Support for nested arrays is already in `ConfigArray`, it's just disabled.
+
+### Introduce an `extends` Key
+
+The `extends` key is an array that may contain other configurations. When used, ESLint will expand the config object into multiple config objects in order to create a flat structure to evaluate. 
+
+#### Extending Objects
+
+You can pass one or more objects in `extends`, like this:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+
+    {
+        files: ["**/src/*.js"],
+        extends: [js.configs.recommended, example.configs.recommended]
+    }
+
+];
+```
+
+Assuming these config objects do not contain `files` or `ignores`, ESLint will convert this config into the following:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+    {
+        ...js.configs.recommended
+        files: ["**/src/*.js"],
+    },
+    {
+        ...example.configs.recommended
+        files: ["**/src/*.js"],
+    },
+];
+```
+
+If the objects in `extends` contain `files`, then ESLint will intersect those values (AND operation); if the object in `extends` contains `ignores`, then ESLint will merge those values (OR operation). For example:
+
+```js
+import globals from "globals";
+
+const config1 = {
+    name: "config1",
+    files: ["**/*.cjs.js"],
+    languageOptions: {
+        sourceType: "commonjs",
+        globals: {
+            ...globals.node
+        }
+    }
+};
+
+const config2 = {
+    name: "config2",
+    files: ["**/*.js"],
+    ignores: ["**/tests"],
+    rules: {
+        "no-console": "error"
+    }
+};
+
+
+export default [
+
+    {
+        name: "myconfig",
+        files: ["**/src/*.js"],
+        ignores: ["**/*.test.js"]
+        extends: [config1, config2],
+        rules: {
+            semi: "error"
+        }
+    }
+
+];
+```
+
+Here, the `files` keys will be combined and the `ignores` key will be merged, resulting in a final config that looks like this:
+
+```js
+
+export default [
+
+    // intersected files and original ignores
+    {
+        name: "myconfig > config1",
+        files: [["**/src/*.js", "**/*.cjs.js"]],
```

Comment:
What would be `files` in the final config when `files` in the user config and extended config have multiple items? For example:

```js
{
    files: ["src/**", "lib/**"],
    extends: [{ files: ["**/*.js", "**/*.mjs"] }]
}
```

Reply:
In order to get the intersection correct, it would look like this:

```js
{
    files: [["src/**", "**/*.js"], ["lib/**", "**/*.js"], ["src/**", "**/*.mjs"], ["lib/**", "**/*.mjs"]]
}
```

Reply:
@nzakas In the scenario you describe, I'd expect this intersection:

```js
{
  files: [["src/**"],   ["**/*.js"]]
  //      ("src/**") && ("**/*.js")
}
```

Which still doesn't match `tests/foo.js`.

I think users and plugin developers can understand this pretty easily. If you provide one array, input files must match any pattern (union). And if you provide multiple arrays, input files must match every array (intersection).

Maybe there's something I'm missing, but I think an intersection of unions will work better than a union of intersections for most cases.

Reply:
> I think users and plugin developers can understand this pretty easily. If you provide one array, input files must match any pattern (union). If you provide multiple arrays, input files must match any pattern (union) in every array (intersection).

If I'm struggling to understand this, I'm not so sure how easily everyone will. :smile:

Ultimately, @mdjermanovic is correct, this is an implementation detail and won't be visible to users.